.travis.yml

```diff
@@ -9,107 +10,59 @@ install:
-      before_script:
-        - pip install --upgrade numpy
-        - pip install --upgrade pandas
-        - pip install --upgrade scikit-learn
-        - pip install --upgrade dask
-        - pip install --upgrade distributed
-        - pip install --upgrade scipy
-        - pip list
-      script:
-        - sed -i 's/\-n auto/\-n 2/g' setup.cfg
-        - "if [ $TRAVIS_PULL_REQUEST == false ] && ! [ $TRAVIS_BRANCH == 'master' ]; then pytest tests/units -n2; else pytest tests; fi"
+      script: *test
+    - env: NUMPY="latest", PANDAS="latest", SCIKITLEARN="latest", DASK="latest", DISTRIBUTED="latest", SCIPY="latest"
+      # pandas 0.25 requires python >= 3.5.3
+      python: 3.5.3
+      script: *test
+    - env: NUMPY="1.15.1", PANDAS="0.23.2", SCIKITLEARN="0.19.2", DASK="0.16.1", DISTRIBUTED="1.18.3", SCIPY="1.2.0"
       python: 3.7
-      after_success:
-        - coveralls
-
-    - stage: Run tests on oldest set of dependencies (Python 3.5.3)
-      env: NUMPY="1.12.0", PANDAS="0.20.3", SCIKITLEARN="0.19.0", DASK="0.15.2", DISTRIBUTED="1.18.3", SCIPY="1.2.0"
-      before_script:
-        - pip install numpy==1.12.0  # First version with official python 3.6 support.
-        - pip install pandas==0.20.3
-        - pip install scikit-learn==0.19.0
-        - pip install dask==0.15.2
-        - pip install distributed==1.18.3
-        # need to downgrade tornado manually
-        - pip install tornado==4.5.3
-        - pip install scipy==1.2.0
-        - pip list
-      script:
-        - sed -i 's/\-n auto/\-n 2/g' setup.cfg
-        - "if [ $TRAVIS_PULL_REQUEST == false ] && ! [ $TRAVIS_BRANCH == 'master' ]; then pytest tests/units -n2; else pytest tests; fi"
+      script: *test
+    - env: NUMPY="1.12.0", PANDAS="0.20.3", SCIKITLEARN="0.19.0", DASK="0.15.2", DISTRIBUTED="1.18.3", SCIPY="1.2.0"
+      python: 3.6
+      script: *test
+    - env: NUMPY="1.12.0", PANDAS="0.20.3", SCIKITLEARN="0.19.0", DASK="0.15.2", DISTRIBUTED="1.18.3", SCIPY="1.2.0"
       python: 3.5.3
+      script: *test
 
-    - stage: Run tests on oldest set of dependencies (Python 3.6)
-      env: NUMPY="1.12.0", PANDAS="0.20.3", SCIKITLEARN="0.19.0", DASK="0.15.2", DISTRIBUTED="1.18.3", SCIPY="1.2.0"
-      before_script:
-        - pip install numpy==1.12.0  # First version with official python 3.6 support.
-        - pip install pandas==0.20.3
-        - pip install scikit-learn==0.19.0
-        - pip install dask==0.15.2
-        - pip install distributed==1.18.3
-        # need to downgrade tornado manually
-        - pip install tornado==4.5.3
-        - pip install scipy==1.2.0
-        - pip list
-      script:
-        - sed -i 's/\-n auto/\-n 2/g' setup.cfg
-        - "if [ $TRAVIS_PULL_REQUEST == false ] && ! [ $TRAVIS_BRANCH == 'master' ]; then pytest tests/units -n2; else pytest tests; fi"
-      python: 3.6
+    - stage: Test coverage
+      if: env(PYTEST_ADDOPTS) = "-c setup.cfg" AND fork = false
+      script: coveralls
```

Comment:
Maybe I am just too stupid, but why is it not possible to do a "normal" python version matrix (as e.g. https://docs.travis-ci.com/user/languages/python/) with the additional `env` variables, just define the test script once (as the script) and have the coveralls as an `after_success` as described here: 
https://github.com/z4r/python-coveralls?
Also, why did you disable it on forks? Sorry for the maybe naive questions

Reply:
The problem is that environment variables depend on Python version; my feeling is that this script would be less readable with a matrix, hence I did not try, but I may be wrong.

About forks, I did not want to make noise when making experiments in my repository, I am not sure whether this is a good idea of not.